federation/planner.go

```diff
@@ -0,0 +1,318 @@
+				if hasField {
+					serviceWithField = service
+				}
+			}
+
+			selectionsByService[serviceWithField] = append(
+				selectionsByService[serviceWithField], selection)
+		}
+	}
+
+	// Create a plan for all the selections that can be resolved in the current graphql service
+	for _, selection := range localSelections {
+		field := typ.Fields[selection.Name]
+		var childPlan *Plan
+		if selection.SelectionSet != nil {
+			var err error
+			childPlan, err = e.plan(field.Type, selection.SelectionSet, service)
+			if err != nil {
+				return nil, fmt.Errorf("planning for %s: %v", selection.Name, err)
+			}
+		}
+
+		selectionCopy := &graphql.Selection{
+			Alias:        selection.Alias,
+			Name:         selection.Name,
+			Args:         selection.Args,
+			SelectionSet: selection.SelectionSet,
+			UnparsedArgs: selection.UnparsedArgs,
+			ParentType:   selection.ParentType,
+		}
+		p.SelectionSet.Selections = append(p.SelectionSet.Selections, selectionCopy)
+
+		if childPlan != nil {
+			for _, subPlan := range childPlan.After {
+				subPlan.Path = append(subPlan.Path, PathStep{Kind: KindField, Name: selection.Alias})
+				p.After = append(p.After, subPlan)
+			}
+		}
+	}
+
+	// needKey is true for selections on other graphql servers
+	needKey := false
+
+	// List of services with selections in the query
+	var otherServices []string
+	for other := range selectionsByService {
+		otherServices = append(otherServices, other)
+	}
+	sort.Strings(otherServices)
+
+	// Create a plan for all selections that can be resolved in other graphql queries
+	for _, other := range otherServices {
+		selections := selectionsByService[other]
+		needKey = true
+
+		subPlan, err := e.plan(typ, &graphql.SelectionSet{Selections: selections}, other)
+		if err != nil {
+			return nil, fmt.Errorf("planning for %s: %v", other, err)
+		}
+
+		p.After = append(p.After, subPlan)
+	}
+
+	// If the selection set doesn't have "__federation" key add it to the selection set
+	// "__federation" indicates a seperate subplan that will be dispatched to a graphql server
+	if needKey {
+		hasKey := false
+		for _, selection := range p.SelectionSet.Selections {
+			if selection.Name == "__federation" && selection.Alias == "__federation" {
+				hasKey = true
+			} else if selection.Name == "__federation" || selection.Alias == "__federation" {
+				return nil, fmt.Errorf("Both the selection name and alias have to be __federation")
+			}
+		}
+		if !hasKey {
+			p.SelectionSet.Selections = append(p.SelectionSet.Selections, &graphql.Selection{
+				Name:  "__federation",
+				Alias: "__federation",
+				Args:  map[string]interface{}{},
+			})
+		}
+	}
+
+	return p, nil
+
+}
+
+func (e *Planner) planUnion(typ *graphql.Union, selectionSet *graphql.SelectionSet, service string) (*Plan, error) {
+	plan := &Plan{
+		// TODO: only include __typename if needed for dispatching? ie. len(types) > 1 and len(fragments) > 0?
+		// TODO: ensure __typename doesn't conflict with another field?
```

Comment:
Are these TODOs for future PRs?  Or should be resolved before merging this PR?

Reply:
yup, was planning to do it in a future PR